### DISCUSSION TITLE
RSTP to MJPEG for use with Octoprint

### DISCUSSION THREAD

**Discussion by StreckerCM** [bluenviron/mediamtx#927] [Questions]
I have been working on this for a few days and I have made some progress but I am having an issue figuring out how to output the RTSP stream as an MJPEG stream that I can use with Octoprint.  Natively Octoprint uses MJPG-Streamer, FFmpeg, and the mpeg2video codec to generate a stream from the hardware camera, but I cannot find any instructions for MJPG-Streamer that use an RSTP input.

This is what I have so far.  I am building a custom rtsp-simple-server container that has FFmpeg and Nginx and a few other utilities that I am using to edit or debug while I figure things out.

```
# syntax=docker/dockerfile:1
FROM aler9/rtsp-simple-server AS rtsp
FROM alpine:3.12

# make sure the package repository is up to date and install apps
RUN apk update \
 && apk upgrade \
 && apk add --no-cache bash \
 && apk add --no-cache nginx \
 && apk add --no-cache ffmpeg \
 && apk add --no-cache v4l-utils \
 && apk add --no-cache nano \
 && apk add --no-cache procps \
 && rm -rf /var/cache/*/* \
 && echo "" > /root/.ash_history \
 && mkdir -p /run/nginx

RUN apk add --no-cache ffmpeg

# change default shell from ash to bash
RUN sed -i -e "s/bin\/ash/bin\/bash/" /etc/passwd

ENV LC_ALL=en_US.UTF-8

# Copy rtsp-simple-server files
COPY --from=rtsp /rtsp-simple-server /
COPY --from=rtsp /rtsp-simple-server.yml /

# Copy Nginx files
#COPY ./video.js /usr/share/nginx/html/
COPY ./nginx.default.conf /etc/nginx/http.d/default.conf
COPY ./nginx.default.conf /etc/nginx/conf.d/default.conf

# Copy Scripts
#COPY ./nginxReloader.sh /usr/local/bin/nginxReloader.sh
#COPY ./docker-entrypoint.sh /usr/local/bin/docker-entrypoint.sh

#RUN chmod +x /usr/local/bin/nginxReloader.sh
#RUN chmod +x /usr/local/bin/docker-entrypoint.sh

ENTRYPOINT ["/rtsp-simple-server"]
```

I have the following paths in configured my rtsp-simple-server.yml file. 

```
paths:
  all:
  direct:
    source: rtsp://liveview:7GQLQsC4D28LrwR8c@192.168.1.219:554/cam/realmonitor>

  thumbnail:
    runOnInit: ffmpeg -i rtsp://localhost:$RTSP_PORT/direct -vframes 1 -y /usr/>
    runOnInitRestart: yes
```

I can get a snapshot from the thumbnail code but I cannot get a stream.  Ideally, I would like to be able to use a similar format to what Ocotrint does where adding _?action=stream_ will give you the MJPEG stream and _?action=snapshot_ would return a snapshot.  Worst case I can use the thumbnail for a snapshot if that isn't possible.

**Comment by aler9** [comment]
Hello, rtsp-simple-server is only able to serve RTSP streams. MJPEG support is not a priority since it doesn't play nice with HD streams and it would require decoding and re-encoding streams, something that we don't want to do in order to preserve performance.

If you want to generate an MJPEG stream from a RTSP stream you can use the defunct (but still useful) ffserver in this way: https://stackoverflow.com/questions/68368511/how-to-convert-rtsp-stream-to-mjpeg-http-using-ffmpeg

[a comment by an automated account is not shown]
